base/reflection.jl

```diff
@@ -73,25 +73,28 @@ function fullname(m::Module)
 end
 
 """
-    names(x::Module; all::Bool = false, imported::Bool = false)
+    names(x::Module; qualified::Bool=true, all::Bool = false, imported::Bool = false)
 
-Get an array of the names exported by a `Module`, excluding deprecated names.
-If `all` is true, then the list also includes non-exported names defined in the module,
+Get an array of the public names of `Module`, excluding deprecated names.
+
+If `qualified` is false, then return only exported names, not public but unexported names.
+If `all` is true, then the list also includes private names defined in the module,
 deprecated names, and compiler-generated names.
 If `imported` is true, then names explicitly imported from other modules
 are also included.
 
-As a special case, all names defined in `Main` are considered \"exported\",
-since it is not idiomatic to explicitly export names from `Main`.
+As a special case, all names defined in `Main` are considered \"public\",
+since it is not idiomatic to mark names from `Main` as public.
 
 See also: [`@locals`](@ref Base.@locals), [`@__MODULE__`](@ref).
 """
-names(m::Module; all::Bool = false, imported::Bool = false) =
-    sort!(unsorted_names(m; all, imported))
-unsorted_names(m::Module; all::Bool = false, imported::Bool = false) =
-    ccall(:jl_module_names, Array{Symbol,1}, (Any, Cint, Cint), m, all, imported)
+names(m::Module; qualified::Bool = true, all::Bool = false, imported::Bool = false) =
+    sort!(unsorted_names(m; qualified, all, imported))
+unsorted_names(m::Module; qualified::Bool = true, all::Bool = false, imported::Bool = false) =
```

Comment:
Should we throw if `qualified=false` and `all=true`? These keywords seem poorly named and inconsistent. Let's change this.

Reply:
"If qualified is false, then return only exported names, not public but unexported names."

only exported names that are unexported names?

Reply:
I have a possible solution:
```
"""
    names(x::Module; visibility = :public, deprecated = false, hidden = false, imported = false)

Get a vector of the names of `Module`, filtered by the keyword arguments.

- `visibility` can be one of `:exported`, `:public`, or `:all`.
   - If `visibility` is `:public` then internal names are filtered out
   - If `visibility` is `:exported` then all unexported names are filtered out.
- If `deprecated` is `false`, then deprecated names are filtered out.
- If `hidden` is `false`, then hidden (e.g. compiler-generated) names are filtered out.
- If `imported` is `false`, then internal names that are imported from other modules are filtered out.

Setting `visibility` to `:all` or any of the other keywords to `true` disables the
corresponding filter. The result is the intersection of the names that pass all the enabled
filters.

As a special case, all names defined in `Main` are considered \"public\",
since it is not idiomatic to mark names from `Main` as public.

For backwards compatibility, it is also possible to pass `all=true` which is equivalent to
passing `visibility=:all`, `deprecated=true`, and `hidden=true` (but not `imported=true`).

!!! Compat 1.11
    The `visibility`, `deprecated`, and `hidden` keyword arguments were added in Julia 1.11.
    To support prior versions of julia, use the `all` keyword argument instead.

See also: [`@locals`](@ref Base.@locals), [`@__MODULE__`](@ref).
"""
```
This docstring should fully describe the proposed behavior and so is ready for review. If the docstring doesn't make it completely clear what I mean, then it's not a good enough docstring. 

I'm holding back on implementing this until I know what [`jl_atomic_load_relaxed(&b->owner) == b`](https://github.com/JuliaLang/julia/blob/0eb13d70a3ea2419f96fd5db5676db91a54ede32/src/module.c#L950) means.

Reply:
I don't think `hidden` is documented elsewhere and I don't know what "compiler-generated names" refers to.

Reply:
Edited to explain `hidden` a bit more.

tbh I'm not entirely sure what the deal with compiler-generated names is. They typically start with an `#`, and they are an implementation detail, so maybe we shouldn't expose them at all. Because compiler-generated names are an implementation detail, it is technically not a violation of symver to make `all=true` no longer return them because their very existence was never guaranteed to continue.

Reply:
I don't think calling this "hidden" is a good idea, if we want to expose this at all. It just seems like another easy-to-confuse thing, thinking of "object hiding" in other languages.

Reply:
I've never heard of "object hiding", would you provide a link please?
I called this `hidden` because that is the variable name used on the c side already.
If you would prefer a different name, feel free to recommend one.
All words have multiple existing meanings, and most applicable words have existing meanings in other programming languages. Because all translation is lossy, those meanings will never perfectly align with the meaning of a word in Julia. For example, the concept of ["data hiding"](https://www.simplilearn.com/tutorials/cpp-tutorial/data-hiding-in-cpp) in c++ is similar to, though not identical to this usage of the word hiding. The main difference is that "data hiding" refers to hiding data, specifically, and this case, which is not prefixed with "data", is not data specific. This case also follows the Julian model of making things harder to reach rather than prohibiting access altogether. Examining c++'s usage of the term "data hiding" makes me more rather than less inclined to call this concept "hiding" here.
It is possible to not expose this at all, but that comes at the cost of not covering all of the existing functionality provided by the `all` keyword.

Reply:
Apologies for not using the exact terminology. I was indeed referring to "data hiding" or "information hiding". I used "object hiding" here because in Julia, this has the effect of generally hiding any object. The semantic meaning is the same. It's precisely this closeness to the existing concept that's the reason for my objection.